Makefile

```diff
@@ -169,27 +81,83 @@ tidyup:
 	@go get -u ./...
 	@go mod tidy
 
-container:
-	@echo "[+] Building container image"
-	@env GOOS=linux CGO_ENABLED=0 go build --ldflags '-s -w -extldflags "-static"' -o $(EXEC_DIR)tcpingDocker $(SOURCE_FILES)
-	@docker build -t tcping:latest .
-	@rm $(EXEC_DIR)tcpingDocker
-	@echo "[+] Done"
-
-gitHubActions:
-	@echo "[+] Building container image - GitHub Actions"
-	@env GOOS=linux CGO_ENABLED=0 go build --ldflags '-s -w -extldflags "-static"' -o tcpingDocker $(SOURCE_FILES)
-	@echo "[+] Done"
```

Comment:
Hmm. I understand why this is okay (everything is statically-linked) but it's not really in the spirit of containerisation, because we're doing everything in an isolated environment. A multi-staged Dockerfile [like this](https://www.saybackend.com/blog/02-golang-dockerfile/) is more idiomatic IMO.

Also the `gitHubActions` rule isn't actually building an image? Not sure what shenanigans are going on here.

That being said, I think these changes would belong in a separate PR for cleanliness' sake.

Reply:
the `gitHubActions` is indeed a leftover. 

also, not sure what we gain here by a multi-stage Dockerfile.

Reply:
> also, not sure what we gain here by a multi-stage Dockerfile.

Currently, the build process itself is not performed in a containerised environment, which goes against the key spirits of containerisation - platform-independence and reproducibility. In short, the build process in its current state is dependent on a working toolchain on the host, which alongside its other environmental factors is largely opaque to the end user (or even to the packager themselves). Setting up the build environment in a containerised environment improves the observability, reliability, and trustworthiness of the build artefact(s).

As of how a multi-stage Dockerfile fits into this equation, this is just so that we don't ship a massive image that unnecessarily contains the toolchain to the user. The build happens in `imageA`; then we copy the relevant artefacts into `imageB` and ship it. It's very well supported at this point and not at all complicated to set up.